pkg/server/handler.go

```diff
@@ -102,6 +104,54 @@ func WithAuditAnnotation(handler http.Handler) http.HandlerFunc {
 	})
 }
 
+// WithVirtualWorkspacesProxy proxies internal requests to virtual workspaces (i.e., requests that did
+// not go through the front proxy) to the external virtual workspaces server. Proxying is required to avoid
+// certificate verification errors because these requests typically come from the kcp loopback client, and it is
+// impossible to use that client against any server other than kcp.
+func WithVirtualWorkspacesProxy(apiHandler http.Handler, shardVirtualWorkspaceURL *url.URL, transport http.RoundTripper) http.HandlerFunc {
+	return func(w http.ResponseWriter, req *http.Request) {
+		logger := klog.FromContext(req.Context())
+
+		if !strings.HasPrefix(req.URL.Path, "/services/") {
+			apiHandler.ServeHTTP(w, req)
+			return
+		}
+
+		if shardVirtualWorkspaceURL == nil || transport == nil {
+			// This handler func is only installed when these are both set. If this happens, it means we've regressed
+			// in the installation of this handler func, and a panic is appropriate.
+			panic("both shardVirtualWorkspaceURL and transport are required")
+		}
+
+		user, ok := request.UserFrom(req.Context())
+		if !ok {
+			http.Error(w, "no user", http.StatusBadRequest)
```

Comment:
Let's use the k8s error writers so clients get a `*metav1.Status`

Reply:
Done